src/awkward1/_util.py

```diff
@@ -1145,6 +1150,65 @@ def recursively_apply(layout, getfunction, args=(), depth=1, keep_parameters=Tru
         )
 
 
+def recursive_walk(layout, apply, args=(), depth=1, materialize=False):
```

Comment:
This is for determining the overall `ptr_lib` from a tree structure of nested array nodes. I split the recursive part and the `ptr_lib`-determining part, with the recursion here in `_util.py` as a utility function. That way, the logic of how to walk down a tree of nested array nodes is in one place, and if we ever add a new node type, it can be updated in one place.

To do that, the action of finding `ptr_lib` has to be defined in a function, with that function being passed in here. The `braodcast_and_apply` and `apply_recursively` work the same way. This is an example of the value of functional programming. There are old articles about how LISP seemed "magical" to C programmers back in the 80's—this kind of factorization over arbitrary functions is very useful! (Though if it's overused, it leads to Javascript's callback hell.)